[Feature] 커스텀 앨범 화면 구현 및 클린아키텍처 적용

letusHyun opened the pull request:
🚨 **Your checklist for this pull request** 

- [x]  Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). Don't request your master!

- [x]  Check the commit's or even all commits' message styles matches our requested structure.

## 📌 [요약]
- 커스텀 앨범 화면 구현
- Domain Layer, Presentation Layer 구현
- PHAsset을 UIImage 혹은 Data로 변환해주는 PhotoService를 Common으로 정의
## ✨ [작업 내용]
### 화면 설명
커스텀 앨범 화면을 구현했습니다. 
카메라 버튼을 누르면 앨범화면을 띄웁니다. 이때 사진을 최대 20장까지 선택할 수 있으며(리뷰 작성 전체 사진 max는 50장으로 고려중입니다), 완료버튼을 누르면 리뷰작성 포스팅란에 선택했던 이미지들이 순서대로 들어갑니다.


<img src="https://github.com/IF-TG/iOS/assets/108918481/6dbb3aeb-a21a-4144-9868-afc3294e5760" width="350">|<img src="https://github.com/IF-TG/iOS/assets/108918481/ca682451-4188-49b1-8fd2-62adc859cfbc" width="350">
|:-:|:-:|
|`처음 앨범 화면`|`사진 3장 클릭한 앨범 화면`|


<center>

| <img src="https://github.com/IF-TG/iOS/assets/108918481/ddf35fc2-20da-4b38-93c9-ed24fb9a220b" width="350"> |
|:-:|
| `앨범사진 cell` |

</center>

- 사진cell을 클릭했을 때, 1,2,3,4사분면 중 1사분면을 클릭하면 앨범 사진이 클릭된 순서를 표시해줍니다. 그리고, 2,3,4 사분면을 클릭하면 사진을 확대하는 화면을 띄울 예정입니다.

<br/><br/><br/>
https://github.com/IF-TG/iOS/assets/108918481/a7abbe05-549c-4414-bad6-094f3c0e5a58

- 사진 선택은 최대 개수는 20개입니다.
- 다음은 이미 선택한 사진을 다시 터치한 경우입니다. 현재 누른 순서를 제거하고, 이후로 밀려있는 순서를 앞당기는 기능입니다. 영상을 보면 이해가 더 빠를 것 같아요.

### UseCase
- usecase를 정의하는데 시간이 좀 걸렸습니다.. 정답이 없어서 더 어려웠던 것 같습니다. Actor는 UseCase를 이용하는 주체이며, usecase를 이용해서 Entity를 얻어냅니다. 또한 UseCase는 비즈니스 로직이 변경되지 않는 한 변경이 없어야 하며, 다른 layer에 의존하지 않아야 합니다. 이를 기반으로 UseCase를 다음과 같이 정의했습니다.
![image](https://github.com/IF-TG/iOS/assets/108918481/91be6ec8-e4d0-4250-8b46-9d86b576caf7) 
	- 앨범으로부터 사진의 정보가 담긴 메타데이터 PHAsset을 fetch
	- 사진 최대 개수 제한 관리하기

참고로 해당 Scene은 API 혹은 localDB를 사용하지 않기 때문에, Repository가 존재하지 않습니다.
PHAsset은 이미지, 동영상 등의 메타데이터가 됩니다. PHAsset과 부가Option을 이용해서 PHImageManager 또는 PHCachingImageManager에게 비동기 요청을 하면, 결과 핸들러를 통해 UIImage 혹은 Data를 가져올 수 있습니다. (저는 서브클래스인 PHCachingImageManager를 사용했는데, 이 객체를 사용하면 자동 캐싱처리까지 해줘서 이미지 스크롤 시 캐싱효과를 볼 수 있습니다)

현재 클린아키텍처를 적용중이라, 레이어가 많아져서 각 레이어마다 이미지를 어떻게 정의하고 들고있어야할지 고민이 많았습니다. 더욱이 ViewModel은 View에 의존하면 안되기 때문에 UIImage형태로 들고있으면 안됩니다. 즉, PHAsset 또는 Data 타입으로 들고 있어야합니다. 따라서 초기에 ViewModel에서 Data타입을 들고있도록 구현을합니다.

### 마주한 문제와 해결
코드 흐름은 이러했습니다. usecase는 사용자의 갤러리 사진 개수만큼 [PHAsset]을 만들어서 ViewModel에게 전달합니다. (정확하게는 usecase가 viewModel을 모르기 때문에 전달한다는 표현은 잘못된 표현이지만 편의상 전달한다고 가정하겠습니다!) 
실제 디바이스로 실행을 해봤는데, 제 갤러리에 사진이 약 1200장 정도가 있었기 때문에 1200개의 PHAsset가 넘어왔고, viewModel에서 이를 Data로 변환하는 작업을 진행했습니다. 이 과정에서 메모리를 초과사용해서 앱이 죽는 문제가 있었습니다.
처음엔 원인을 몰라 헤매다가 instruments을 통해 확인해봤더니 메모리를 약 2.5GB를 사용하고 있었습니다.
![image](https://github.com/IF-TG/iOS/assets/108918481/98b85677-5e0b-4ebc-ba79-6faa8aded580)
그래서 cellForItemAt시점에 dequeue된 cell들에 한에서 UIImage로 변환하는 코드로 수정을 하게 되었습니다.


![image](https://github.com/IF-TG/iOS/assets/108918481/fceb33a1-4ca8-4221-8da6-c8d85a46f247)

추가로 fetchImage메소드를 사용할 때, Option을 정의해줘야하는데, 이때 resizeMode에 따라서 image의 품질이 달라집니다.
따라서 앨범 화면인 collectionView Cell에서 보여줄 때는 화질을 낮춰서 성능을 개선시켰고, 리뷰작성 화면에서의 이미지는 성능을 낮춰서 화질을 높혔습니다.

![image](https://github.com/IF-TG/iOS/assets/108918481/8df06605-1659-4c48-8180-c14109cf03bf)
![image](https://github.com/IF-TG/iOS/assets/108918481/70944c6f-cec4-4f2a-91cf-478587f6286a)

## 📸  [시연 영상]
*<!-- 스크린샷이 필요하면 스크린샷을 첨부해주세요 -->*
https://github.com/IF-TG/iOS/assets/108918481/a7abbe05-549c-4414-bad6-094f3c0e5a58


## TODO
- 중앙 상단에 최근항목기능 뿐 아니라, 다른 정렬 리스트 추가하기
- 사진 상세 화면 구현하기
## 📚 [레퍼런스 (또는 새로 알게 된 내용) 혹은 궁금한 사항]

*<!-- 참고할 사항이 있다면 적어주세요 -->*



### reference
- https://developer.apple.com/documentation/photokit/phcachingimagemanager
- https://developer.apple.com/documentation/photokit/phimagerequestoptionsresizemode
- https://developer.apple.com/documentation/photokit
- https://ios-development.tistory.com/1415
- https://ios-development.tistory.com/1416
- https://ios-development.tistory.com/1417

letusHyun committed:
[Chore] package update

letusHyun committed:
Merge branch 'feature/login' into feature/album

# Conflicts:
#	travelPlan/travelPlan.xcworkspace/xcshareddata/swiftpm/Package.resolved

letusHyun committed:
[Chore] IDEPreferLogStreaming=YES를 추가해서 Logging Error 해결

letusHyun committed:
[Chore] Logging Error 최종 해결

letusHyun committed:
Merge branch 'develop' of https://github.com/IF-TG/iOS into feature/album

letusHyun committed:
[Chore] package

letusHyun committed:
[Rename] Photo_에서 Album_으로 네이밍 변경

letusHyun committed:
[System] pbxproj

letusHyun committed:
[Chore] package

letusHyun committed:
[Feat] AlbumViewModel, UseCase 추가 및 초기 설정

letusHyun committed:
[System] pbxproj

letusHyun committed:
[Feat] PhotoQuadrant Entity 추가

letusHyun committed:
[Remove] 다시 제거

letusHyun committed:
[Feat] 의존성 주입

letusHyun committed:
[Remove] AlbumService 삭제

해당 파일을 삭제하고, AlbumUsecase가 이 기능을 대체합니다.

letusHyun committed:
[System] instruments사용 시, profile을 debug로 설정

letusHyun committed:
[Rename] Entity로 위치 변경

letusHyun committed:
[Rename] PhotoCell.Quadrant를 PhotoCellQuadrant로 변경

letusHyun committed:
[Rebuild] 삭제한 메소드를 usecase에서 정의하도록 수정

letusHyun committed:
[Feat] 의존성 주입 추가

letusHyun committed:
[Feat] AlbumUseCase 함수 반환타입 변환

letusHyun committed:
[Rebuild] 프로퍼티 수정 및 cellInfo 정의

letusHyun committed:
[Remove] AlbumInfo파일 삭제

letusHyun committed:
[Rebuild] configue 인자 타입 변경

letusHyun committed:
[Rebuild] 네이밍 변경 및 불필요 코드 제거

letusHyun committed:
[System] pbxproj

letusHyun committed:
[Rebuild] PhotoEntity 재정의

letusHyun committed:
[Remove] 해당 로직을 Usecase에서 수행

letusHyun committed:
[Common] PHAsset extension으로 UIImage로 변환하는 함수 추가

letusHyun committed:
[Remove] extension보다 configure 매개변수 주입이 더 나을것 같아 파일 삭제

extenion으로 사용하면 PHAsset을 image로 변환하는 함수 호출 시, 인자로 targetSize, contentMode, options등을 정의해주어야 합니다.
즉, cell 내부에서 targetSize등 여러 파라미터값들을 조절하게됩니다. 이러한 방법은 cell의 재사용서을 떨어뜨릴 수 있을 것 같아서 이를 해결하기 위해 vc에서 photoService를 이용해 image를 fetch해오고,
최종적으로 viewModel에서 PhotoModel을 추가 정의해주었습니다.

letusHyun committed:
[Remove] 불필요 파일 제거

letusHyun committed:
[Rename] PhotoService를 Common Utils 디렉토리로 파일 위치 변경

앨범에서 사진을 받아오는 로직이 여러 화면에서 사용될 수 있기 때문에 변경합니다.

letusHyun committed:
[Feat] AlbumUseCase 최종 구현

letusHyun committed:
[Feat] AlbumViewModel 최종 구현

letusHyun committed:
[Feat] VC 완료버튼 데이터 전달 로직 제외하고 구현

letusHyun committed:
[System] pbxproj

letusHyun committed:
[Style] Cell 자잘한 코드 개선

letusHyun committed:
[Comment] 사진 정렬 주석 추가

letusHyun committed:
[Feat] PHAsset passing 로직 추가

letusHyun committed:
[Style] 앨범에서 가져온 사진의 size 값 Constant로 수정

letusHyun committed:
[Feat] PhotoService 구현체 메소드 내부에서 메인스레드로 변경하는 로직 추가

letusHyun committed:
[Feat] input output 컨벤션 준수

letusHyun committed:
[Feat] 완료 버튼 클릭 시, 선택했던 사진의 메타데이터인 [PHAsset]을 부모 VC에게 전달

letusHyun committed:
Merge branch 'develop' of https://github.com/IF-TG/iOS into feature/album

# Conflicts:
#	travelPlan/travelPlan.xcodeproj/project.pbxproj

letusHyun committed:
[Remove] 불필요 파일 제거

letusHyun committed:
[System] pbxproj

letusHyun committed:
[Style] 불필요 코드 제거

letusHyun committed:
[Style] 불필요 코드 제거

letusHyun committed:
[Feat] 인자에 resizeModeOption 추가

letusHyun committed:
[Feat] 각 화면에 맞게 이미지 품질 조정

letusHyun committed:
[Feat] prepareForReuse 시점에 이미이 nil로 초기화

SHcommit commented on travelPlan/Source/Presentation/Album/View/PhotoCell.swift:
```diff
@@ -90,10 +90,9 @@ final class PhotoCell: UICollectionViewCell {
-    if case let .selected(order) = info?.selectedOrder {
+  func configure(with cellInfo: PhotoCellInfo) {
+    imageView.image = cellInfo.image
+    if case let .selected(order) = cellInfo.selectedOrder {
```
기본적으로 prepareForReuse 시점은 전부 not selecetd UI로 하고 configure(with:)시점에 .selected(order)인 경우에만 UI를 바꾸는 방법은 어떨까요?

letusHyun replied:
오호.. 좋은 방법이네요. 수정했습니다!

SHcommit commented on travelPlan/Source/Presentation/Album/ViewController/AlbumViewController.swift:
```diff
@@ -233,44 +221,19 @@ private extension PhotoViewController {
-      print("연관값으로 image넘겨서 imageDetailVC 열기")
-      // TODO: - 연관값으로 image넘겨서 imageDetailVC 열기
-      break
+extension AlbumViewController: PhotoCellDelegate {
+  func touchBegan(_ cell: UICollectionViewCell, quadrant: PhotoCellQuadrant) {
+    guard let indexPath = collectionView.indexPath(for: cell) else { return }
+    if case .first = quadrant {
+      input.touchedFirstQuadrant.send(indexPath)
+    } else {
+      input.touchedElseQuadrant.send(indexPath)
```
오 뷰컨의 호출 로직이 간결해졌네요+_+

letusHyun replied:
감사합니다!

SHcommit commented on travelPlan/Source/Presentation/Album/ViewController/AlbumViewController.swift:
```diff
@@ -128,36 +124,34 @@ extension PhotoViewController: UICollectionViewDataSource {
-        let cellInfo = PhotoCellInfo(asset: asset, image: image, selectedOrder: imageInfo.selectedOrder)
-        cell?.configure(with: cellInfo)
-      }
+      contentMode: .aspectFit,
+      resizeModeOption: .fast
+    ) { [weak cell] image in
+      let cellInfo = PhotoCellInfo(image: image, selectedOrder: photoModel.selectedOrder)
+      cell?.configure(with: cellInfo)
+    }
+
```
초기 문제를 PHAsset인 메타데이터를 이용해서 화면에 보여질 cell에 대해서만 비동기로 UIimage로 변환 후 화면에 보여지도록 변환하셨네요: ]


저도 나름 해결 방안을 생각해봤는데 페이징을 기반으로 사용자가 스크롤할 때마다 enumerateObjects(at:options:using:) <a href="https://developer.apple.com/documentation/foundation/nsarray/1417577-enumerateobjects">관련 링크</a> 이 함수를 통해 통해 특정 범위에 대해서만 requestImage를 호출해서 이미지 저장하고 indexPath에 따라서 바로바로 비동기 처리가 아니라 동기적으로 반환하는건 어떨까 생각했습니다.
이 경우 추가적으로 PhotoService 객체를 presentation에서 따로 소유하고 있지 않고 유즈케이스에서만 처리할 수 있을 것 같다는 생각이 들었습니다.

그러나 아무래도 PHCachingImageManager를 통해 내부적으로 한번 로컬 앨범에서 받아온 이미지는 비동기처리없이 인메모리에서 가져오기 때문에,, 데이터를 두번이나 VM이나 PhotoService 객체에서 소유할 필요는 없다는 생각이 또 들게되었습니다.

---

화면에 보이는 cell에 대해서만 fetch하는 방법이 좋은 해결 방식인것 같습니다🤩

SHcommit reviewed (commented):
확인했습니다: )

한 가지 궁금한 것은 photoKit에서 제공하는 image resizing된 이미지의 화질, 데이터와 ImageIO에서제공하는 리사이징 방식 중 어느게 더 효율적인지 (단순 궁금하네요),,

letusHyun committed:
[Rebuild] prepareForReuse 시점에 전부 not selecetd UI로 하고 configure(with:)시점에 .selected(order)인 경우에만 UI를 바꾸도록 수정

letusHyun committed:
[Rebuild] 여기서 메인스레드로 바꾸니까 completoin 내부에서 무거운 작업을 하게될 때 문제 발생 할 수 있어서 메인큐 제거

letusHyun committed:
[Rebuild] 이미지가 순서대로 load되지 않는 문제 해결

사용자가 사진을 선택한 순서에 맞게 이미지를 load해야 합니다.
하지만 fetchImage는 내부적으로 비동기호출을 하는데, 이 결과를 view에서 바로 이미지뷰로 만들다보니 순서가 발생했습니다.
이를 해결하기 위해 DispatchGroup을 사용해서 호출하는 순서와 그에 맞는 image를 저장했고 모든 비동기 호출이 완료되면 순서를 기반으로 imageView를 추가했습니다.

letusHyun committed:
Merge branch 'develop' of https://github.com/IF-TG/iOS into feature/album

letusHyun committed:
[Chore] package
